internal/controllers/topology/cluster/structuredmerge/serversidepathhelper_test.go

```diff
@@ -0,0 +1,442 @@
+		g := NewWithT(t)
+
+		// Get the current object (assumes tests to be run in sequence).
+		original := obj.DeepCopy()
+		g.Expect(env.GetAPIReader().Get(ctx, client.ObjectKeyFromObject(original), original)).To(Succeed())
+
+		// Create a patch helper for a modified object with some changes to what previously applied by th topology manager.
+		modified := obj.DeepCopy()
+		g.Expect(unstructured.SetNestedField(modified.Object, "changed", "spec", "controlPlaneEndpoint", "host")).To(Succeed())
+		g.Expect(unstructured.SetNestedField(modified.Object, "changed-by-topology-controller", "spec", "bar")).To(Succeed())
+
+		p0, err := NewServerSidePatchHelper(original, modified, env.GetClient(), IgnorePaths{{"spec", "foo"}})
+		g.Expect(err).ToNot(HaveOccurred())
+		g.Expect(p0.HasChanges()).To(BeTrue())
+		g.Expect(p0.HasSpecChanges()).To(BeTrue())
+
+		// Create the object using server side apply
+		g.Expect(p0.Patch(ctx)).To(Succeed())
+
+		// Check the object and verify the change is applied as well as managed field updated accordingly.
+		got := obj.DeepCopy()
+		g.Expect(env.GetAPIReader().Get(ctx, client.ObjectKeyFromObject(got), got)).To(Succeed())
+
+		v2, _, _ := unstructured.NestedString(got.Object, "spec", "bar")
+		g.Expect(v2).To(Equal("changed-by-topology-controller"))
+
+		fieldV1 := getTopologyManagedFields(got)
+		g.Expect(fieldV1).ToNot(BeEmpty())
+		g.Expect(fieldV1).To(HaveKey("f:spec")) // topology controller should express opinions on spec.
+
+		specFieldV1 := fieldV1["f:spec"].(map[string]interface{})
+		g.Expect(specFieldV1).ToNot(BeEmpty())
+		g.Expect(specFieldV1).To(HaveKey("f:controlPlaneEndpoint")) // topology controller should express opinions on spec.controlPlaneEndpoint.
+		g.Expect(specFieldV1).ToNot(HaveKey("f:foo"))               // topology controller should not express opinions on ignore paths.
+		g.Expect(specFieldV1).To(HaveKey("f:bar"))                  // topology controller now has an opinion on a field previously managed by other controllers (force ownership).
+	})
+	t.Run("No-op on unstructured object having empty map[string]interface in spec", func(t *testing.T) {
+		// TODO: this should be either a unit test or a test with a typed CRD
+		g := NewWithT(t)
+
+		// Second test object having an empty map and slice as value
+		obj2 := builder.InfrastructureClusterTemplate(ns.Name, "obj2").WithSpecFields(map[string]interface{}{
+			"spec.emptyMap":   map[string]interface{}{},
+			"spec.emptySlice": []interface{}{}, // this field is then explicitly ignored by the patch helper
+		}).Build()
+
+		// create new object having an empty map[string]interface in spec and a copy of it for further testing
+		original := obj2.DeepCopy()
+		modified := obj2.DeepCopy()
+
+		// Create the object using server side apply
+		g.Expect(env.PatchAndWait(ctx, original, client.FieldOwner("topology"))).To(Succeed())
```

Comment:
Valid point:
* However for this case (having `manager=topology` the `dryRunPatch` also returns `false, false`.
  * it identifies that it was not managed before (because managed fields for `capi-topology` is empty
  * also: the original and modified object is the same (which means nothing we added got removed)

So in this case it is OK that it does not detect a change.

Setting anything in spec would result in `manager=topology` case to detect a change, but this would be a whole other test-case than `No-op on unstructured object having empty map[string]interface in spec` (because we would have to define something in spec!?

Reply:
this test was giving false results because it was using `InfrastructureCluster`, which uses XpreserveFields true.
Now the test is using `TestInfrastructureCluster`, with a typed CRD

Reply:
Was fixed in https://github.com/kubernetes-sigs/cluster-api/pull/6495/commits/91a23e9d505e06e8c794409378352384dc793896